File: module-android-api/src/main/java/com/tainzhi/sample/api/adapter/FilterableAdapter.java
package com.tainzhi.sample.api.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.tainzhi.sample.api.R;

import java.util.ArrayList;
import java.util.List;

/**
 * @author: tainzhi
 * @mail: qfq61@qq.com
 * @date: 2019-10-13 19:42
 * @description:
 **/

public class FilterableAdapter extends RecyclerView.Adapter<FilterableAdapter.MyViewHolder> implements Filterable {
	private Context mContext;
	private List<String> mList = new ArrayList<>();
	private List<String> mOriginalList = new ArrayList<>();
	
	public FilterableAdapter(Context context, List<String> list) {
		this.mContext = context;
		mList = list;
		mOriginalList = list;
	}
	
	@NonNull
	@Override
	public MyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
		View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_basic_vertial_view, parent,
				false);
		return new MyViewHolder(view);
	}
	
	@Override
	public void onBindViewHolder(@NonNull MyViewHolder holder, int position) {
		holder.textView.setText(mList.get(position));
	}
	
	@Override
	public int getItemCount() {
		return mList.size();
	}
	
	public void removeItem(int position) {
		mList.remove(position);
		mOriginalList.remove(position);
		notifyItemRemoved(position);
	}
	
	public void restoreItem(String item, int position) {
		mList.add(position, item);
		mOriginalList.add(position, item);
		notifyItemInserted(position);
	}
	
	
	@Override
	public Filter getFilter() {
		return new Filter() {
			@Override
			protected FilterResults performFiltering(CharSequence constraint) {
				String constraintString = constraint.toString();
				List<String> filteredList = new ArrayList<>();
				if (constraintString.isEmpty()) {
					filteredList = mOriginalList;
				} else {
					for (String s : mOriginalList) {
						if (s.toLowerCase().contains(constraintString)) {
							filteredList.add(s);
						}
					}
				}
				FilterResults filterResults = new FilterResults();
				filterResults.values = filteredList;
				return filterResults;
			}
			
			@Override
			protected void publishResults(CharSequence constraint, FilterResults results) {
				mList = (ArrayList<String>) results.values;
				notifyDataSetChanged();
			}
		};
	}
	
	public class MyViewHolder extends RecyclerView.ViewHolder {
		public TextView textView;
		public View foreground, background;
		
		public MyViewHolder(View view) {
			super(view);
			textView = view.findViewById(R.id.text_view);
			foreground = view.findViewById(R.id.cl_foreground);
			background = view.findViewById(R.id.cl_background);
		}
	}
}

File: module-android-api/src/main/java/com/tainzhi/sample/api/adapter/FilterableAdapter.kt
package com.tainzhi.sample.api.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.tainzhi.sample.api.R
import java.util.*

/**
 * @author: tainzhi
 * @mail: qfq61@qq.com
 * @date: 2019-10-13 19:42
 * @description:
 */
class FilterableAdapter(
    private val mContext: Context,
    list: MutableList<String>
) : RecyclerView.Adapter<FilterableAdapter.MyViewHolder>(),
    Filterable {
    private var mList: List<String> = ArrayList()
    private var mOriginalList: MutableList<String> =
        ArrayList()

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): MyViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(
            R.layout.item_basic_vertial_view, parent,
            false
        )
        return MyViewHolder(view)
    }

    override fun onBindViewHolder(
        holder: MyViewHolder,
        position: Int
    ) {
        holder.textView.text = mList[position]
    }

    override fun getItemCount(): Int {
        return mList.size
    }

    fun removeItem(position: Int) {
        mList.removeAt(position)
        mOriginalList.removeAt(position)
        notifyItemRemoved(position)
    }

    fun restoreItem(item: String, position: Int) {
        mList.add(position, item)
        mOriginalList.add(position, item)
        notifyItemInserted(position)
    }

    override fun getFilter(): Filter {
        return object : Filter() {
            override fun performFiltering(constraint: CharSequence): FilterResults {
                val constraintString = constraint.toString()
                var filteredList: MutableList<String> =
                    ArrayList()
                if (constraintString.isEmpty()) {
                    filteredList = mOriginalList
                } else {
                    for (s in mOriginalList) {
                        if (s.toLowerCase().contains(constraintString)) {
                            filteredList.add(s)
                        }
                    }
                }
                val filterResults = FilterResults()
                filterResults.values = filteredList
                return filterResults
            }

            override fun publishResults(
                constraint: CharSequence,
                results: FilterResults
            ) {
                mList = results.values as ArrayList<String>
                notifyDataSetChanged()
            }
        }
    }

    inner class MyViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        var textView: TextView
        var foreground: View
        var background: View

        init {
            textView = view.findViewById(R.id.text_view)
            foreground = view.findViewById(R.id.cl_foreground)
            background = view.findViewById(R.id.cl_background)
        }
    }

    init {
        mList = list
        mOriginalList = list
    }
}